File: tests/cross_driver/java/src/main/java/com/secantus/smokes/ListDbFilterSmoke.java
package com.secantus.smokes;

import com.mongodb.ConnectionString;
import com.mongodb.MongoClientSettings;
import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoClients;
import org.bson.Document;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;

/**
 * Cross-driver listDatabases filter smoke — Java.
 * <p>
 * Insert one doc into three databases, then list with
 * {@code filter: {name: "alpha"}} and assert only that one is returned.
 * Run again with {@code nameOnly: true} and assert at least three dbs
 * are listed.
 */
public final class ListDbFilterSmoke {

    private ListDbFilterSmoke() {}

    public static void main(String[] args) {
        String uri = System.getenv("MONGODB_URI");
        if (uri == null) {
            System.err.println("MONGODB_URI not set");
            System.exit(2);
        }

        MongoClientSettings settings = MongoClientSettings.builder()
            .applyConnectionString(new ConnectionString(uri))
            .applyToClusterSettings(b -> b.serverSelectionTimeout(30, TimeUnit.SECONDS))
            .build();

        try (MongoClient client = MongoClients.create(settings)) {
            for (String dbName : List.of("alpha", "beta", "gamma")) {
                client.getDatabase(dbName).getCollection("c").insertOne(new Document("_id", 1));
            }
            try {
                Document filtered = client.getDatabase("admin").runCommand(new Document()
                    .append("listDatabases", 1)
                    .append("filter", new Document("name", "alpha")));
                @SuppressWarnings("unchecked")
                List<Document> dbs = (List<Document>) filtered.get("databases");
                List<String> names = new ArrayList<>();
                if (dbs != null) {
                    for (Document d : dbs) {
                        names.add(d.getString("name"));
                    }
                }
                if (names.size() != 1 || !"alpha".equals(names.get(0))) {
                    System.err.printf("filter: got %s, want [alpha]%n", names);
                    System.exit(1);
                }

                Document namesOnly = client.getDatabase("admin").runCommand(new Document()
                    .append("listDatabases", 1)
                    .append("nameOnly", true));
                @SuppressWarnings("unchecked")
                List<Document> all = (List<Document>) namesOnly.get("databases");
                if (all == null || all.size() < 3) {
                    System.err.printf("nameOnly: got %d dbs, want >= 3%n",
                        all == null ? 0 : all.size());
                    System.exit(1);
                }
                System.out.println("OK");
            } finally {
                for (String dbName : List.of("alpha", "beta", "gamma")) {
                    try {
                        client.getDatabase(dbName).drop();
                    } catch (RuntimeException ignored) {
                        // best-effort cleanup
                    }
                }
            }
        }
    }
}

File: tests/cross_driver/php/listdb_filter_smoke.php
<?php
// Cross-driver listDatabases filter smoke — PHP.
//
// Insert one doc into three databases, then list with `name: alpha`
// filter and assert only that one is returned. Re-list with
// `nameOnly: true` and assert at least three dbs.

declare(strict_types=1);

require __DIR__ . '/vendor/autoload.php';

use MongoDB\Client;
use MongoDB\Driver\Command;

function bail(string $msg): never {
    fwrite(STDERR, $msg . PHP_EOL);
    exit(1);
}

$uri = getenv('MONGODB_URI') ?: bail('MONGODB_URI not set');

$client = new Client($uri, ['serverSelectionTimeoutMS' => 5000]);

foreach (['alpha', 'beta', 'gamma'] as $dbName) {
    $client->selectCollection($dbName, 'c')->insertOne(['_id' => 1]);
}

try {
    $reply = $client->getManager()->executeCommand('admin', new Command([
        'listDatabases' => 1,
        'filter' => ['name' => 'alpha'],
    ]));
    $filtered = (array) $reply->toArray()[0];
    $names = array_map(fn($d) => ((array) $d)['name'], (array) $filtered['databases']);
    if ($names !== ['alpha']) {
        bail('filter: got ' . json_encode($names) . ', want [alpha]');
    }

    $reply2 = $client->getManager()->executeCommand('admin', new Command([
        'listDatabases' => 1,
        'nameOnly' => true,
    ]));
    $all = (array) $reply2->toArray()[0];
    $dbs = (array) $all['databases'];
    if (count($dbs) < 3) {
        bail('nameOnly: got ' . count($dbs) . ' dbs, want >= 3');
    }

    echo "OK" . PHP_EOL;
} finally {
    foreach (['alpha', 'beta', 'gamma'] as $dbName) {
        try { $client->dropDatabase($dbName); } catch (\Throwable $e) { /* best-effort */ }
    }
}
